File: src/aocpython/day21/art.py
from math import sqrt

picture = '.#./..#/###'

class Picture():
    def __init__(self, start, rules):
        self.picture = start.split('/')
        self._rules = rules

    def enhance(self):
        parts = self._parts()
        enhanced_parts = [self._rules.match(part) for part in parts]
        self._assemble(enhanced_parts)

    def count_pixels(self):
        return ''.join(self.picture).count('#')

    def _parts(self):
        size = 2 if len(self.picture) % 2 == 0 else 3
        return [tuple([self.picture[k][j:j+size] for k in range(i, i+size)]) for i in range(0, len(self.picture), size) for j in range(0, len(self.picture), size)]

    def _assemble(self, parts):
        size = int(sqrt(len(parts)))
        self.picture = [row for chunk in [[''.join(row) for row in zip(*parts[j:j+size])] for j in range(0, len(parts), size)] for row in chunk]


class Rules():
    def __init__(self, raw_rules):
        lines = [line.rsplit() for line in raw_rules]
        self._rules = {}
        for line in lines:
            self._rules[tuple(line[0].split('/'))] = tuple(line[2].split('/'))

    def match(self, block):
        matches = [block, self._flip(block)]
        for i in range(3):
            matches += [self._rotate(b) for b in matches[-2:]]
        for m in matches:
            if m in self._rules.keys(): return self._rules[m]

    def _rotate(self, block):
        return tuple([''.join(line) for line in zip(*block[::-1])])

    def _flip(self, block):
        return tuple([line[::-1] for line in block])

with open('./data/21') as puzzle_input:
    rules = Rules(puzzle_input)
    picture = Picture(picture, rules)

    [picture.enhance() for _ in range(5)]
    print(picture.count_pixels())
    [picture.enhance() for _ in range(13)]
    print(picture.count_pixels())

File: src/AocHaskell/Day21/Art.hs
import Data.List (transpose)
import Data.List.Split (splitOn)
import qualified Data.Map.Strict as M
import Data.Foldable (asum)
import qualified Data.Vector as V

data Pixel = On | Off deriving (Eq, Ord)
type Picture = V.Vector (V.Vector Pixel)
type Rules = M.Map Picture Picture

vecTranspose :: V.Vector (V.Vector a) -> V.Vector (V.Vector a)
vecTranspose = V.fromList . map V.fromList . transpose . V.toList . V.map V.toList

disassemble :: Picture -> V.Vector Picture
disassemble picture =
    V.fromList [V.map (V.slice j size) . (V.slice i size) $ picture | i <- range, j <- range]
    where
        size = if length picture `mod` 2 == 0 then 2 else 3
        range = [0, size.. length picture - 1]

assemble :: V.Vector Picture -> Picture
assemble parts = 
    V.concat [V.map V.concat . V.map V.toList . vecTranspose . (V.slice j size) $ parts | j <- [0, size.. length parts - 1]]
    where size = floor . sqrt . fromIntegral . V.length $ parts

enhance :: Rules -> Picture -> Picture
enhance rules picture
    | length picture > 3 = assemble . V.map (enhance rules) . disassemble $ picture
    | otherwise = case asum . map (\p -> M.lookup p rules) $ transforms picture of Just x -> x
    where transforms picture = concat . take 4 . iterate (map rotate) $ [picture, reflect picture]
            where
                reflect = V.map V.reverse
                rotate = reflect . vecTranspose

readPicture :: String -> Picture
readPicture = V.fromList . map (V.fromList . map (\c -> case c of '#' -> On; '.' -> Off)) . splitOn "/" 

countPixels :: Picture -> Int
countPixels = sum . V.map (V.length . V.filter ((==) On))


main = do
    puzzleInput <- readFile "./data/21"
    let rules = M.fromList . map (\rule -> (readPicture $ rule !! 0, readPicture $ rule !! 2)). map words . lines $ puzzleInput
    let enhancements =  take 19 $ iterate (enhance rules) (readPicture ".#./..#/###")
    print $ map (countPixels . (!!) enhancements) [5, 18]